content/en/BestPractices/defend-against-ssrf.md

```diff
@@ -0,0 +1,257 @@
+
+public ResponseEntity Resource> getPackagel
+
+        @Pathvariable String group,
+
+        @PathVariable String artifact,
+
+        @PathVariable String version) {
+
+    Path pkgPath = Paths.get (String.join("/", group.split("\\.")))
+
+            .resolve(artifact)
+
+            . resolve (version)
+
+            . resolve (String. format ("%S-s.jar", artifact, version))
+
+            . normalize() ;
+
+    URI pkgUrl = BASE _URI.resolve (pkgPath.toString());
+
+    return sendResponseStream(this.okHttpWrapper.get(pkgUrl.toString()));
+
+}
+
+```
+
+It is recommended to analyze the methods used to handle user-supplied input for any known security concerns and avoid using them. In case there is no alternative, writing a wrapper around it is recommended for further filtration and preventing an attack from execution. 
+
+---
+
+### Scenario - 3: Extracting Cloud Metadata 
+
+
+
+Suppose the application is powered by Cloud services such as AWS, Digital Ocean, Google Cloud, etc... In that case, An attacker may attempt to leverage an SSRF issue to steal the cloud metadata credentials. 
+
+
+
+Let's understand this with a practical code example: 
+
+
+
+```
+
+<?php
+
+
+
+# Check if the 'url' GET variable is set
+
+if (isset($_GET['image_url'])){
+
+$url = $_GET['image_url'];
+
+
+
+# Fetch the image from the user-supplied URL
+
+$file = fopen($url, 'rb');
+
+
+
+# Send proper header for png images
+
+header("Content-Type: image/png");
+
+
+
+# Dump image file
+
+fpassthru($file);
+
+}
+
+
+
+echo 'Please enter image url'
+
+
+
+?>
+
+```
+
+
+
+In the above code, the application accepts user input from the `image_url` parameter, which is passed to the `$url` function without any sanitization. The `header("Content-Type: image/png");` is being checked for `image/png` header type. Lastly, the `$file` passes malicious user-supplied input to the `fpassthru()` function and returns the data to the user. 
+
+
+
+Now, an attacker could attempt to extract metadata by crafting the following malicious payload: 
+
+
+
+- http://127.0.0.1/ssrf.php?image_url=http://169.254.169.254/latest/meta-data/hostname
+
+
+
+This payload `http://169.254.169.254/latest/meta-data/hostname` will return the `hostname` related information. Similarly, an attacker could extract `AWS Secret`, which could be used to gain privileged access to the AWS account and compromise it. 
+
+
+
+In the case of AWS Cloud, attacks on cloud metadata could be mitigated by applying the proper validation on the code level and using a solution like IMDSv2. 
+
+---
+
+Since we have now looked at various exploitation scenarios to gain a thorough understanding of SSRF vulnerabilities, it is crucial to learn how we can mitigate and prevent SSRF attacks. 
+
+
+## Prevention Measures 
+
+
+**1. Disallow unnecessary protocols:** It is recommended to restrict the use of unnecessary protocols such as `file://, gopher://, schema://` and others as they could be used to bypass the restrictions. 
+
```

Comment:
For consistency, we don't usually bold the Numbers in reported, but the Head Words ("Disallow unnecessary protocols", e.g) can be bold, and work nicely that way.

Reply:
I'd also cut "it is recommended" and just start with "Restrict"